### ISSUE TITLE
Problems related to WordPress PHPUnit tests

### ISSUE THREAD

**Issue by roberthusak** [peachpiecompiler/peachpie#791]
This issue summarizes all the problems which cause WordPress PHPUnit tests to end prematurely (uncaught .NET exception, stack overflow etc.) or pollute output with extra text. Currently, they are solved by workarounds - by commenting certain parts out or by skipping. Some of them might require further investigation.

- [ ] `function_exists( 'add_filter' )` ([see here](https://github.com/WordPress/wordpress-develop/blob/master/tests/phpunit/includes/functions.php#L55)) is replaced by `true` even though `add_filter` (defined in `PeachPied.WordPress`) might not be defined in the current `Context` if it's definition hasn't been included yet.

- [ ] `symlink` function is defined but throws `NotSupportedException`, which is not caught by PHPUnit.

- [ ] `imagecopyresampled` doesn't support negative `$src_w` and `$src_h`. They should flip the image, [see here](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/class-wp-image-editor-gd.php#L377).

- [ ] `DOMDocument::loadHTML` throws an exception when parsing `/` in `<input readonly/>` - it thinks that `readonly/` is a (obviously invalid) name of a parameter.

- [ ] `XSLTProcessor::importStylesheet` doesn't support `simplexml`.

- [ ] `getid3_quicktime.QuicktimeParseAtom` ([see here](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/ID3/module.audio-video.quicktime.php#L230)) occupies 267 200 bytes on the call stack, causing stack overflow.

- [ ] When `$x` is e.g. `stdClass`, `$x[0]` causes `InvalidOperationException`, unable to be caught by PHPUnit.

- [ ] It seems that the MySQL connections are not closed (maybe due to a single `Context` for all tests and different behaviour of destructors?), reaching default number of `max_connections` option in about the third of the tests (returning the error: `Too many connections`). The current workaround is to increase the option manually.

- [ ] `file_exists` doesn't take into account source files contained in PHAR archives.

- [ ] As `wp-includes/class-json.php` is excluded from WordPress compilation due to the missing `PEAR_Error` class, the test of `Services_JSON` must be skipped for now.

- [ ] Class `GP_Locale` is dynamically downloaded and included ([see here](https://github.com/WordPress/wordpress-develop/blob/master/tests/phpunit/tests/pomo/pluralForms.php#L52)), which currently causes its definition to be output. Including it in the tests statically should prevent this from happening.

**Comment by jakubmisek** [comment]
@roberthusak can we implement `symlink` using `Mono.Unix.UnixSymbolicLinkInfo.CreateSymbolicLinkTo()` ?

**Comment by jakubmisek** [comment]
@roberthusak I think `$x[0]` causes a fatal error in PHP which won't be handled either? Isn't it a different bug in PeachPie?

**Comment by roberthusak** [comment]
> @roberthusak can we implement `symlink` using `Mono.Unix.UnixSymbolicLinkInfo.CreateSymbolicLinkTo()` ?

@jakubmisek That's true, we can make it work the same way as `chown`. If we want to make it run on Windows as well (PHP does that since Windows Vista), we'll just need to find an appropriate way (maybe using P/Invoke as [here](https://stackoverflow.com/a/11156870/2105235)), as .NET Standard seems not to have the API [yet](https://github.com/dotnet/runtime/issues/24271).

> @roberthusak I think `$x[0]` causes a fatal error in PHP which won't be handled either? Isn't it a different bug in PeachPie?

In fact, it throws a catchable `Error` (it seems the fatal error is caused by not catching it), try this code (tested on PHP 7.4.1):

```php
function test($a) {
  try {
    echo $a[0];
  } catch (Throwable $e) {
    echo get_class($e);
  }
}

test(new stdClass());
```

**Comment by jakubmisek** [comment]
Note - large `getid3_quicktime.QuicktimeParseAtom` stack size might be caused when debugging since the compiler emits `IndirectLocal` for each local variable to be displayed in IDE's "Locals" window.

**Comment by roberthusak** [comment]
> Note - large `getid3_quicktime.QuicktimeParseAtom` stack size might be caused when debugging since the compiler emits `IndirectLocal` for each local variable to be displayed in IDE's "Locals" window.
> 
> This might help: [df2266a](https://github.com/peachpiecompiler/peachpie/commit/df2266aa88ef0f0ef0a235fc97e55a00df34c975)

Unfortunately, this doesn't help. The stack overflow still happens even with the latest commit of Peachpie and even in `Release`.

**Comment by jakubmisek** [comment]
issue with stack size in `getid3_quicktime.QuicktimeParseAtom` is related to https://github.com/dotnet/runtime/issues/8980

The method is too large for JIT. The solution would be for us to emit fewer temp value types on the stack.
